src/rp_data_tree.c

```diff
@@ -0,0 +1,476 @@
+ */
+
+#include "rp_data_tree.h"
+#include "xpath_processor.h"
+#include "sr_common.h"
+
+
+/**
+ * @brief Creates part of xpath for leaf and container nodes. Arguments specify if the namespace and trailing slash
+ * should be included.
+ */
+static int
+rp_dt_create_xpath_for_cont_leaf_node(const struct lyd_node *data_tree, char **xpath, bool namespace, bool trailing_slash)
+{
+    CHECK_NULL_ARG2(data_tree, xpath);
+    char *s = NULL;
+    size_t len = 1; /* terminating null byte*/
+    size_t ns_len = 0;
+    size_t node_len = 0;
+
+    /* calculate length */
+    if (namespace){
+        ns_len = strlen(data_tree->schema->module->name) + 1; /*namespace + colon*/
+        len += ns_len;
+    }
+    node_len = strlen(data_tree->schema->name);
+    len += node_len;
+    if (trailing_slash){
+        len++;
+    }
+    s = calloc(sizeof(char),len);
+    if (NULL == s){
+        SR_LOG_ERR_MSG("Memory allocation failed");
+        return SR_ERR_NOMEM;
+    }
+
+    /* copy string */
+    if(namespace){
+        strcpy(s, data_tree->schema->module->name);
+        s[ns_len-1] = ':';
+    }
+    strcpy(s+ns_len, data_tree->schema->name);
+
+    if(trailing_slash){
+        s[ns_len+node_len] = '/';
+    }
+
+    *xpath = s;
+    return SR_ERR_OK;
+}
+
+/**
+ * @brief Creates part of xpath for list nodes. Arguments specify if the namespace and trailing slash
+ * should be included.
+ */
+static int
+rp_dt_create_xpath_for_list_node(const struct lyd_node *data_tree, char **xpath, bool namespace, bool trailing_slash)
+{
+    CHECK_NULL_ARG2(data_tree, xpath);
+    char *s = NULL;
+    size_t len = 1; /* terminating null byte*/
+    size_t ns_len = 0;
+    size_t offset = 0;
+
+    /* calculate length*/
+    if (namespace) {
+        ns_len = strlen(data_tree->schema->module->name);
+        len += ns_len + 1; /*namespace + colon*/
+    }
+    len += strlen(data_tree->schema->name);
+    if (trailing_slash) {
+        len++;
+    }
+    /* lookup keys */
+    struct lys_node_list *sch_list = (struct lys_node_list *) data_tree->schema;
+
+    struct lyd_node_leaf_list **key_nodes = calloc(sch_list->keys_size, sizeof(struct lyd_node_leaf_list *));
+    if (key_nodes == NULL){
+        SR_LOG_ERR_MSG("Memory allocation failed");
+        return SR_ERR_NOMEM;
+    }
+    struct lyd_node *c = data_tree->child;
+    size_t matched = 0;
+
+    while (c != NULL){
+        for (int k=0; k < sch_list->keys_size; k++){
+            if(0 == strcmp(sch_list->keys[k]->name, c->schema->name)){
+                key_nodes[matched] = (struct lyd_node_leaf_list *) c;
+                len += strlen(key_nodes[matched]->value_str); /*key value*/
+                len += strlen(c->schema->name); /*key name*/
+                len += 5; /*delimiting characters [='']*/
+                matched++;
+                break;
+            }
+        }
+        c = c->next;
+    }
+    if (matched != sch_list->keys_size){
+        SR_LOG_ERR_MSG("Keys not found");
```

Comment:
Be more verbose by error messages so that we triage the issue better from logs, e.g. also print the node name.